apps/fipsinstall.c

```diff
@@ -0,0 +1,397 @@
+      "fips.conf -mac HMAC -macopt digest:SHA256 -macopt hexkey:00"},
+    {"in", OPT_IN, 's', "Input module file name (with no extension)"},
+    {"cfg", OPT_CFG, '>', "config file name to output or verify"},
+    {"mac", OPT_MAC, 's', "MAC name"},
+    {"macopt", OPT_MACOPT, 's', "MAC algorithm control parameters in n:v form. "
+                                "See 'Supported Controls' in the EVP_MAC_ docs"},
+    {"verify", OPT_VERIFY, '-', "Verify the installed status is correct"},
+    {"section", OPT_SECTION_NAME, 's', "the section name in the config file"},
+
+    {NULL}
+};
+
+static int mac_ctrl_string(EVP_MAC_CTX *ctx, const char *value)
+{
+    int rv;
+    char *stmp, *vtmp = NULL;
+
+    stmp = OPENSSL_strdup(value);
+    if (stmp == NULL)
+        return -1;
+    vtmp = strchr(stmp, ':');
+    if (vtmp != NULL) {
+        *vtmp = 0;
+        vtmp++;
+    }
+    rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+    OPENSSL_free(stmp);
+    return rv;
+}
+
+static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
+                  unsigned char *out, size_t *out_len)
+{
+    int ret = 0;
+    int i;
+
+    if (!EVP_MAC_init(ctx))
+        goto err;
+    if (EVP_MAC_size(ctx) > *out_len)
+        goto end;
+    for (;;) {
+        i = BIO_read(in, (char *)tmp, BUFSIZE);
+        if (i < 0)
+            goto err;
+        if (i == 0)
+            break;
+        if (!EVP_MAC_update(ctx, tmp, i))
+            goto err;
+    }
+end:
+    if (!EVP_MAC_final(ctx, out, out_len))
+        goto err;
+    ret = 1;
+err:
+    return ret;
+}
+
+static int hexstr2bio(BIO *out, const unsigned char *data, size_t len)
+{
+    size_t i;
+
+    if (len == 0)
+        return 0;
+
+    for (i = 0; i < len; i++) {
+        BIO_printf(out, "%02X", data[i]);
+    }
+    return 1;
+}
+
+static int print_mac(BIO *bio, const char *label, const unsigned char *mac,
+                     size_t len)
+{
+    if (!BIO_printf(bio, "%s =", label))
+        return 0;
+    if (!hexstr2bio(bio, mac, len))
+        return 0;
+    return BIO_printf(bio, "\n");
+}
+
+static int load_fips_prov_and_run_self_test(char *fname,
+                                            const unsigned char *mac,
+                                            size_t mac_len)
+{
+    char buffer[1024], *buf;
+    size_t sz;
+    int ret = 0;
+    OSSL_PROVIDER *prov = NULL;
+    char *prov_name, *dot;
+    static OSSL_PARAM params[2];
+    OSSL_PARAM param_end = OSSL_PARAM_END;
+    BIO *mem_bio = NULL;
+
+    mem_bio = BIO_new_mem_buf(buffer, sizeof(buffer));
+    if (mem_bio == NULL)
+        return 0;
+
+    if (!hexstr2bio(mem_bio, mac, mac_len))
+        goto end;
+
+    params[0] = OSSL_PARAM_construct_utf8_ptr(OSSL_PROV_PARAM_MODULE_MAC,
+                                              &buf, &sz);
+    params[1] = param_end;
+    OSSL_PARAM_set_utf8_ptr(&params[0], buffer);
+
+    /* strip the path and extension off the name */
```

Comment:
Hmm, the user should be able to register a known module as a provider with a name of their own choice...  I see that the public OSSL_PROVIDER API is lacking in this regard.  Hang on, PR coming up!

(side note: the documentation of `OPT_IN` says "no extension", so I don't think you should try to strip it off here)

Reply:
I need some way of getting the full path of the module name before the provider is loaded so that I can calculate the mac of the fips module file and then pass that to the provider_load. I am currently doing this by passing it into the install application as a command line parameter. That is the reason why it strips out the provider name. I guess it really needs to load the name from the config file (if that is possible before the provider loads).

Reply:
That won't work quite right...  The provider name is meant to be exactly that, a name.  With the public API, there's only one way to affect the location of the module, and that's with the environment variable `OPENSSL_MODULES`, so the *risk* you're taking here is that while this application reads a file `/whatever/blah.so`, the provider functionality tries to load `/official/path/blah.so`...  you do see how that'll fail, right?  That strengthens my though that constructing an internal "config file" in a buffer might be the best path forward...

Reply:
I thought the module path was able to be specified in the config file? And I am trying to do the mac calculation inside the app.. I guess it could move into the core and just be triggered to be calculated before loading the fips module?? So the config entry passed from the app would just be a trigger item that is not normally set maybe?
I was also trying to do a verify (which also used the mac  - which might be a bit harder (unless it gets passed back to the app (sigh))

Reply:
ideally the app would calculate the mac as it is the one responsible for writing out the config file.

Reply:
Through the config file, yes.  Through `OSSL_PROVIDER_load`, no.

So how about this?  Create a buffer that contains a full configuration:

```
openssl_conf = openssl_init

[openssl_init]
{name} = {section}

[{section}]
module = {module_path}
module-mac = {module_mac}
install-version = {install-version}
install-mac = {install-mac}
install-status = {install-status}
```

All things within braces are things for you to generate.

Then make a mem_buf BIO with that buffer (come to think of it, you can construct that buffer via that same mem_buf BIO) and feed the BIO to `CONF_modules_load`.  Voilà, the provider and its extra parameters are stored.  At this point, `OSSL_PROVIDER_load(NULL, name)` makes sense...

Reply:
Regarding provider *name* vs *module name*, you might want to consider making that two different program arguments, i.e.:

```
    -name           Name of the provider
    -module path    File name of the provider module (optional)
```

`-module` would make sense to use if the provider module has a different name, or is located in a non-default location.

Reply:
Hmm So there is another real config file sitting somewhere that could possibly have a different module path.. Sounds not good to me... Would we be better off loading the real config file and tweaking it in memory?? I do need access to the same module path or bad things could happen.

Reply:
I fail to see why the exact module path is important for self testing, or even its name.  Both of these are really application / libcrypto side data, and as long as that data can be passed to the fips provider, it should be fine, no?  I.e. it should be possible to move `fips.so` and `fips.conf` from `/whatever/blah` to `/something/better` and it should still validate, shouldn't it?

Reply:
BTW, something I can't quite make out is how you want the section you provide to be used.  There are two options:

1.  You write a complete section for the module, i.e. something like this:

    ```
    [{section}]
    identity = {name}
    module = {module_path}    # optional, e.g. if the module is in a non-default location
    module-mac = {module_mac}
    install-version = {install_version}
    install-mac = {install_mac}
    install-status = {install_status}
    ```

    and then, it gets used by the system config like this:

    ```
    openssl_conf = openssl_init

    [openssl_init]
    providers = provider_section

    [provider_section]
    {name} = {section}

    .include /path/to/fips.conf
    ```

2.  You write a separate section for the FIPS specific stuff, like this:

    ```
    [{section}]
    module-mac = {module_mac}
    install-version = {install_version}
    install-mac = {install_mac}
    install-status = {install_status}
    ```

    and then, it gets used by the system config like this:

    ```
    openssl_conf = openssl_init

    [openssl_init]
    providers = provider_section

    [provider_section]
    {name} = {name}_section

    [{name}_section]
    identity = {name}       # not really necessary
    module = {module_path}    # optional, e.g. if the module is in a non-default location
    fips-data = {section}

    .include /path/to/fips.conf
    ```

Thing is, the way I see you currently writing things, it looks like you want to *write* a FIPS config file like in 2, but want to be able to *use* a FIPS config file as in 1...  Either way is fine, but you must decide!  This impacts the names of the parameters that the provider has access to:

1.  ```
    identity
    module
    module-mac
    install-version
    install-mac
    install-status
    ```

2.  ```
    identity
    module
    fips-data.module-mac
    fips-data.install-version
    fips-data.install-mac
    fips-data.install-status
    ```

Do you understand the difference?

Reply:
Yes I was aware of the difference.
Looks like using the dummy config will work.